compiler/optimizer/OMRSimplifierHandlers.cpp

```diff
@@ -2220,6 +2220,11 @@ static bool simplifyISelectCompare(TR::Node *compare, TR::Simplifier *s)
       TR::NodeChecklist safetyVisited(s->comp());
       TR_ComparisonTypes compareType = TR::ILOpCode::getCompareType(compare->getOpCodeValue());
       bool isUnsignedCompare = TR::ILOpCode(compare->getOpCode()).isUnsignedCompare();
+
+      // do not try to simplify unsigned compare when it will yield a BadILOp with TR_cmpNE as its TR_ComparisonTypes
+      if (isUnsignedCompare && compare->getFirstChild()->getDataType() != TR::Address)
```

Comment:
This relies on knowledge that the [implementation of `OMR::ILOpCode::compareOpCode`](https://github.com/eclipse-omr/omr/blob/f43ed4cb2ce6fe825717ae52539d303978b8d3c3/compiler/il/OMRILOps.cpp#L98-L154) will return `TR::BadILOp` for unsigned `TR_cmpNE` with integer data types, and only in that case.  I would suggest instead calling `TR::ILOpCode::compareOpCode` first and testing whether the result is `TR::BadILOp`.  That would allow the simplification to apply if the implementation of `OMR::ILOpCode::compareOpCode` ever changes to return a comparison opcode for that case.

Reply:
I did a little bit of research into the history of this.  It looks like `TR::iucmpeq`, _et al._ were removed after the various unsigned arithmetic operations were removed, as they were no longer considered to be needed - comparing two values for equality or inequality will yield that same result regardless of whether the integer values being compared are treated as signed or unsigned quantities.

Given that this transformation is rewriting the `select` so that it will yield either zero or one, and then compares that result to zero, maybe it would be better to remove the `isUnsignedCompare` argument from the call to `::ILOpCode::compareOpCode` below.  That would force that final comparison to always be a signed comparison with zero.

Reply:
Thanks, @hzongaro. That makes sense about the signed comparison being valid for equal/not-equal. I think having the check for BadILOp as you rightly point out to guard against any future implementation changes in `compareOpCode()` might still be a good idea, and we could also remove the `isUnsignedCompare` argument from the call to `compareOpCode()` so that we shouldn't in fact be getting a BadILOp.

Reply:
I've made this change in the recent push.